### ISSUE TITLE
[rllib] PyTorch PPO / APPO

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#3365]
### Describe the problem
Currently we only have a simple pytorch example for A3C. It should be possible to "port" a bunch of pytorch algorithms onto the TorchPolicyGraph class and get them basically for free.

We'll also have to improve pytorch support in some ways (e.g., dict space support, lstm).

**Comment by ericl** [comment]
cc @michaeltu1 @megankawakami

**Comment by nilsjohanbjorck** [comment]
Dear Eric,

I might be interested in taking a stab at implementing either APE-X or IMPALA in pytorch. For APE-X I've looked into rllib.dqn.dqn_policy_graph.py to see what interface it provides. Based upon this class, it seems like I'd need to subclass TorchPolicyGraph and then implement the following functions which provide the main parts of the algorithm:

```
gradients(self, optimizer)
postprocess_trajectory(self,sample_batch,other_agent_batches=None,episode=None)
compute_td_error(self, obs_t, act_t, rew_t, obs_tp1, done_mask, importance_weights)
```

The functions below are also provided in dqn_policy_graph.py, although their implementation seems straightforward

```
optimizer(self)
extra_compute_action_feed_dict(self)
extra_compute_grad_fetches(self)
update_target(self)
set_epsilon(self, epsilon)
get_state(self)
set_state(self, state)
```

Is there anything else that would be needed?

**Comment by ericl** [comment]
Hey @nilsjohanbjorck,

Looking at TorchPolicyGraph, it seems you'd also have to replace the compute_actions() method, which currently assumes that the model returns are logits and does a softmax instead of the argmax / eps. greedy exploration for Q learning: https://github.com/ray-project/ray/blob/9266909c9dbf2e9abc7ec1e23e2a6a26185128b0/python/ray/rllib/evaluation/torch_policy_graph.py#L74

Also, a bunch of features like returning td_error from compute_gradients() are just missing in TorchPolicyGraph (you can see here the extra returns are hardcoded to `{}` https://github.com/ray-project/ray/blob/9266909c9dbf2e9abc7ec1e23e2a6a26185128b0/python/ray/rllib/evaluation/torch_policy_graph.py#L89) so you'd need to get that to work too.

**Comment by nilsjohanbjorck** [comment]
@ericl,

Sorry for the late reply, I'm at a conference. Overriding the compute_actions() method of TorchPolicyGraph seems straightforward. What is needed from the compute_gradients() method seems to be describer here

https://github.com/ray-project/ray/blob/9266909c9dbf2e9abc7ec1e23e2a6a26185128b0/python/ray/rllib/optimizers/async_replay_optimizer.py#L39-L41

The class DQNPolicyGraph overrides the method ```extra_compute_grad_fetches(self) ``` which provides the td-information. That function and ```extra_compute_action_feed_dict(self)```seems to be unique to the class TFPolicyGraph and I assume these wouldn't need to be implemented in a pytorch version. Apart from providing td-information in ```compute_gradients(self, postprocessed_batch)```,  are there any additional hard-coded ```{}```s that would need to be filled out? It seems to not be the case to me, but I'm not that familiar with the code-base.

**Comment by ericl** [comment]
Hm yeah, that might be it. I would just try to see what makes sense to
implement DQN properly, and not worry too much about what the TF version is
doing. Then making the change to get it working in Ape-X should be pretty
straightforward.

**Comment by nilsjohanbjorck** [comment]
Sorry on my inactivity on this issue, I haven't been able to work on it lately. I hope to get back to it eventually, but I cannot really say when I'll be able to. Sorry about that!

**Comment by sven1977** [comment]
Torch PPO is learning. Waiting for 2 other PR-dependencies to get solved first before we can merge this one.
APPO next.

**Comment by sven1977** [comment]
I'm closing this issue.
- PyTorch PPO has been added and tested some time ago (w/ version 0.8.3).
- PyTorch APPO is on its way into the master branch via this PR here: https://github.com/ray-project/ray/pull/8120
